File: backend/src/ytforge/interfaces/api/schemas/channels.py
from __future__ import annotations

import uuid
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from ytforge.domain.enums import ChannelRole


class ChannelCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    youtube_channel_id: str | None = None
    brand_kit: dict[str, Any] = Field(default_factory=dict)
    defaults: dict[str, Any] = Field(default_factory=dict)


class ChannelRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    youtube_channel_id: str | None
    brand_kit: dict[str, Any]
    defaults: dict[str, Any]


class ChannelMemberAddRequest(BaseModel):
    user_id: uuid.UUID
    role: ChannelRole


class ChannelMemberRoleUpdateRequest(BaseModel):
    role: ChannelRole


class ChannelMemberRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    channel_id: uuid.UUID
    user_id: uuid.UUID
    role: ChannelRole

File: frontend/src/lib/api/schemas/channels.ts
import { z } from "zod";
import { ChannelRole } from "@/types/enums";

const channelRoleSchema = z.enum([
  ChannelRole.OWNER,
  ChannelRole.ADMIN,
  ChannelRole.EDITOR,
  ChannelRole.VIEWER,
]);

export const channelReadSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  youtube_channel_id: z.string().nullable(),
  brand_kit: z.record(z.string(), z.unknown()),
  defaults: z.record(z.string(), z.unknown()),
});
export type ChannelRead = z.infer<typeof channelReadSchema>;

export const channelMemberReadSchema = z.object({
  id: z.string().uuid(),
  channel_id: z.string().uuid(),
  user_id: z.string().uuid(),
  role: channelRoleSchema,
});
export type ChannelMemberRead = z.infer<typeof channelMemberReadSchema>;

export type ChannelCreateRequest = {
  name: string;
  youtube_channel_id?: string | null;
  brand_kit?: Record<string, unknown>;
  defaults?: Record<string, unknown>;
};

export type ChannelMemberAddRequest = {
  user_id: string;
  role: ChannelRole;
};

export type ChannelMemberRoleUpdateRequest = {
  role: ChannelRole;
};
